parlai/chat_service/services/websocket/agents.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import time
+import logging
+from queue import Queue
+
+from parlai.core.agents import Agent
+
+
+class WebsocketAgent(Agent):
+    """Class for a person that can act in a ParlAI world via websockets"""
+
+    def __init__(self, opt, manager, task_id, socketID):
+        super().__init__(opt)
+        self.manager = manager
+        self.id = socketID
+
+        self.active = True
+        self.msg_queue = Queue()
+        self.stored_data = {}
+        self.task_id = task_id
+        self.message_request_time = None
+        self.set_stored_data()
+
+    def set_stored_data(self):
+        """Gets agent state data from manager"""
+        agent_state = self.manager.get_agent_state(self.id)
+        if agent_state is not None and hasattr(agent_state, 'stored_data'):
+            self.stored_data = agent_state.stored_data
+
+    def observe(self, act):
+        """Send an agent a message through the websocket manager"""
+        logging.info(f"Sending new message: {act}")
+        if 'payload' in act:
+            raise ValueError("Payload not supported yet by websockets")
+        else:
+            msg = act['text']
+            self.manager.observe_message(self.id, msg)
+
+    def put_data(self, message):
+        """Put data into the message queue if it hasn't already been seen"""
+        logging.info(f"Received new message: {message}")
+        action = {'episode_done': False, 'text': message}
+        self.msg_queue.put(action)
+
+    def get_new_act_message(self):
+        """Get a new act message if one exists, return None otherwise"""
+        # Check if person has sent a message
+        if not self.msg_queue.empty():
+            return self.msg_queue.get()
+
+        return None
+
+    def act(self, timeout=None):
+        """Pulls a message from the message queue. If none exist returns None
+        unless the timeout has expired.
+        """
+        # if this is the first act since last sent message start timing
+        if self.message_request_time is None:
+            self.message_request_time = time.time()
+
+        # If checking timeouts
+        if timeout:
+            # If time is exceeded, timeout
+            if time.time() - self.message_request_time > timeout:
+                return self.mark_inactive()
```

Comment:
should this be defined somewhere?

Reply:
Defined it as a `pass` as in the messenger implementation. Included this in the PR description as a function that can be generalized out.